src/WebJobs.Script.WebHost/Security/BlobStorageSecretsRepository.cs

```diff
@@ -0,0 +1,163 @@
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost
+{
+    /// <summary>
+    /// An <see cref="ISecretsRepository"/> implementation that uses the file system as the backing store.
+    /// </summary>
+    public sealed class BlobStorageSecretsRepository : ISecretsRepository, IDisposable
+    {
+        private readonly string _secretsFileMarkersPath;
+        private readonly string _secretsBlobPath;
+        private readonly string _hostSecretsMarkerPath;
+        private readonly string _hostSecretsBlobPath;
+        private readonly AutoRecoveringFileSystemWatcher _fileWatcher;
+        private readonly CloudBlobContainer _blobContainer;
+        private readonly string _secretsContainerName = "azure-webjobs-secrets";
+        private readonly string _accountConnectionString;
+        private bool _disposed = false;        
+
+        public BlobStorageSecretsRepository(string secretMarkersDirectoryPath, string accountConnectionString, string siteHostName)
+        {
+            if (secretMarkersDirectoryPath == null)
+            {
+                throw new ArgumentNullException(nameof(secretMarkersDirectoryPath));
+            }
+            if (accountConnectionString == null)
+            {
+                throw new ArgumentNullException(nameof(accountConnectionString));
+            }
+
+            _secretsFileMarkersPath = secretMarkersDirectoryPath;
+            _hostSecretsMarkerPath = Path.Combine(_secretsFileMarkersPath, ScriptConstants.HostMetadataFileName);            
+
+            Directory.CreateDirectory(_secretsFileMarkersPath);
+
+            _fileWatcher = new AutoRecoveringFileSystemWatcher(_secretsFileMarkersPath, "*.json");
+            _fileWatcher.Changed += OnChanged;
+
+            _secretsBlobPath = siteHostName.ToLowerInvariant(); // Config.ScriptSettingsManager.Instance.AzureWebsiteDefaultSubdomain;
+            _hostSecretsBlobPath = string.Format("{0}/{1}", _secretsBlobPath, ScriptConstants.HostMetadataFileName);
+
+            _accountConnectionString = accountConnectionString;
+            CloudStorageAccount account = CloudStorageAccount.Parse(_accountConnectionString);
+            CloudBlobClient client = account.CreateCloudBlobClient();
+
+            _blobContainer = client.GetContainerReference(_secretsContainerName);
+            _blobContainer.CreateIfNotExists();
+            //await _blobContainer.CreateIfNotExistsAsync();        
+        }
+
+        public event EventHandler<SecretsChangedEventArgs> SecretsChanged;
+
+        private string GetSecretsBlobFilePath(ScriptSecretsType secretsType, string functionName = null)
+        {
+            return secretsType == ScriptSecretsType.Host
+                ? _hostSecretsBlobPath
+                : string.Format("{0}/{1}", _secretsBlobPath, string.Format(CultureInfo.InvariantCulture, "{0}.json", functionName.ToLowerInvariant()));
+        }
+
+        private string GetSecretsFilePath(ScriptSecretsType secretsType, string functionName = null)
```

Comment:
In this case, rename this to `GetSecretsMarkerFilePath` or something along those lines. I thing we want to consider using the term `sentinel` instead of marker as it is what is already used in other places in the runtime.